src/backend/distributed/operations/citus_tools.c

```diff
@@ -93,6 +101,95 @@ CheckConnectionToNode(char *nodeName, uint32 nodePort)
 }
 
 
+/*
+ * citus_check_cluster_node_health UDF performs connectivity checks from all the nodes to
+ * all the nodes, and report success status
+ */
+Datum
+citus_check_cluster_node_health(PG_FUNCTION_ARGS)
+{
+	CheckCitusVersion(ERROR);
+
+	TupleDesc tupleDescriptor = NULL;
+	Tuplestorestate *tupleStore = SetupTuplestore(fcinfo, &tupleDescriptor);
+
+	StoreAllConnectivityChecks(tupleStore, tupleDescriptor);
+
+	/* clean up and return the tuplestore */
+	tuplestore_donestoring(tupleStore);
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * GetConnectivityCheckCommand returns the command to check connections to a node
+ */
+static char *
+GetConnectivityCheckCommand(const char *nodeName, const uint32 nodePort)
+{
+	StringInfo connectivityCheckCommand = makeStringInfo();
+	appendStringInfo(connectivityCheckCommand,
+					 "SELECT citus_check_connection_to_node('%s', %d)",
+					 nodeName, nodePort);
+
+	return connectivityCheckCommand->data;
+}
+
+
+/*
+ * StoreAllConnectivityChecks performs connectivity checks from all the nodes to all the
+ * nodes, and report success status
+ */
+static void
+StoreAllConnectivityChecks(Tuplestorestate *tupleStore, TupleDesc tupleDescriptor)
+{
+	Datum values[CONNECTIVITY_CHECK_COLUMNS];
+	bool isNulls[CONNECTIVITY_CHECK_COLUMNS];
+
+	List *workerNodeList = ActivePrimaryNodeList(ShareLock);
+
+	WorkerNode *workerNode = NULL;
+	foreach_ptr(workerNode, workerNodeList)
+	{
+		const char *nodeName = workerNode->workerName;
+		const int nodePort = workerNode->workerPort;
+		int32 connectionFlags = 0;
+
+		MultiConnection *connection =
+			StartNodeConnection(connectionFlags, nodeName, nodePort);
```

Comment:
Thanks for this explanation. I wonder if it makes sense to print a better warning message if we haven't finished connection establishment.

In theory that should never happen. However, it can also help us find bugs and make development easier. What do you think @onderkalaci ?

Reply:
I think a good improvement could be to make `StartNodeConnection ` a static function at least.

And, yes, we could consider add some sanity checks if you use a connection that is not successfully established yet.

But, probably in a separate PR